compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs

```diff
@@ -1835,6 +1836,75 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         }
     }
 
+    pub(super) fn collect_unused_for_coerce_return_ty(
+        &self,
+        errors: &Vec<traits::FulfillmentError<'tcx>>,
+    ) {
+        struct SemiVisitor<'a, 'tcx> {
+            fcx: &'a FnCtxt<'a, 'tcx>,
+            semi_expr: Option<&'tcx hir::Expr<'tcx>>,
+            fail_span: Span,
+            return_ty: Ty<'tcx>,
+        }
+
+        impl<'a, 'tcx> Visitor<'tcx> for SemiVisitor<'a, 'tcx> {
+            fn visit_stmt(&mut self, s: &'tcx Stmt<'tcx>) {
+                match s.kind {
+                    hir::StmtKind::Semi(ref expr) => {
+                        if !matches!(expr.kind, hir::ExprKind::Ret(..))
+                            && expr.span.contains(self.fail_span)
+                        {
+                            let expr_ty = self.fcx.typeck_results.borrow().expr_ty(expr);
+                            let expr_ty = self.fcx.tcx.erase_regions(expr_ty);
+                            if self.fcx.can_coerce(expr_ty, self.return_ty) {
+                                self.semi_expr = Some(expr);
+                            }
+                        }
+                    }
+                    _ => {}
+                }
+                intravisit::walk_stmt(self, s);
+            }
+        }
```

Comment:
Is it possible to avoid a visitor by looking at the obligation cause code of the fulfillment error's root obligation? It should very likely be an `ExprBindingObligation`.